lib/Backend/PreLowerPeeps.cpp

```diff
@@ -28,6 +35,139 @@ IR::Instr *Lowerer::PreLowerPeepInstr(IR::Instr *instr, IR::Instr **pInstrPrev)
+{
+    if (!opndFold->GetIsDead())
+    {
+        return instrAdd;
+    }
+
+    if (!(opndAdd->IsRegOpnd() || (opndAdd->IsInt32() && opndAdd->IsIntConstOpnd())))
+    {
+        return instrAdd;
+    }
+
+    if (!opndFold->IsRegOpnd() || !opndFold->AsRegOpnd()->m_sym->IsSingleDef())
+    {
+        return instrAdd;
+    }
+
+    IR::Instr * instrDef = opndFold->AsRegOpnd()->m_sym->GetInstrDef();
+
+    if ((instrDef->m_opcode != Js::OpCode::Shl_I4 && instrDef->m_opcode != Js::OpCode::Mul_I4) || !instrDef->GetSrc1()->IsRegOpnd()  || !instrDef->GetSrc2()->IsIntConstOpnd())
+    {
+        return instrAdd;
+    }
+
+    if (instrDef->HasBailOutInfo())
+    {
+        return instrAdd;
+    }
+
+    byte scale = 0;
+    IntConstType constVal = instrDef->GetSrc2()->AsIntConstOpnd()->GetValue();
+    if (instrDef->m_opcode == Js::OpCode::Shl_I4)
+    {
+        if (constVal < 0 || constVal > 3)
+        {
+            return instrAdd;
+        }
+        scale = (byte)constVal;
+    }
+    else
+    {
+        Assert(instrDef->m_opcode == Js::OpCode::Mul_I4);
+        switch (constVal)
+        {
+        case 1:
+            scale = 0;
+            break;
+        case 2:
+            scale = 1;
+            break;
+        case 4:
+            scale = 2;
+            break;
+        case 8:
+            scale = 3;
+            break;
+        default:
+            return instrAdd;
+        }
+    }
+
+    FOREACH_INSTR_IN_RANGE(instrIter, instrDef->m_next, instrAdd->m_prev)
+    {
+        // if any branch between def-use, don't do peeps on it because branch target might depend on the def
+        if (instrIter->IsBranchInstr())
+        {
+            return instrAdd;
+        }
+        if (instrIter->HasBailOutInfo())
+        {
+            return instrAdd;
+        }
+        if (instrIter->FindRegDef(instrDef->GetSrc1()->AsRegOpnd()->m_sym))
+        {
+            return instrAdd;
+        }
+        if (instrIter->FindRegUse(opndFold->AsRegOpnd()->m_sym))
+        {
+            return instrAdd;
+        }
+    } NEXT_INSTR_IN_RANGE;
+
+    IR::IndirOpnd * leaOpnd = nullptr;
+    if (opndAdd->IsInt32() && opndAdd->IsIntConstOpnd())
+    {
+        leaOpnd = IR::IndirOpnd::New(instrDef->UnlinkSrc1()->AsRegOpnd(), opndAdd->AsIntConstOpnd()->AsInt32(), scale, opndFold->GetType(), instrAdd->m_func);
+    }
+    else
+    {
+        Assert(opndAdd->IsRegOpnd());
+        leaOpnd = IR::IndirOpnd::New(opndAdd->AsRegOpnd(), instrDef->UnlinkSrc1()->AsRegOpnd(), scale, opndFold->GetType(), instrAdd->m_func);
+    }
+
+    IR::Instr * leaInstr = InsertLea(instrAdd->UnlinkDst()->AsRegOpnd(), leaOpnd, instrAdd);
+
+    instrAdd->Remove();
+    instrDef->Remove();
+
+    return leaInstr;
+}
+
+IR::Instr *
+Lowerer::PeepShiftAdd(IR::Instr *instrAdd)
+{
+    // Peep:
+    //    t1 = SHL X, 0|1|2|3    / t1 = MUL X, 1|2|4|8
+    //    t2 = ADD t1, Y
+    //
+    // Into:
+    //    t2 = LEA [X * scale + Y]
+
```

Comment:
@MikeHolman I'm really sorry to bother you, but I was wondering if you could point me to where we double check that there are no subsequent uses of `t1` **after** `t2 = ... ` ?